### ISSUE TITLE
golint asks you to drop else block containing return, even when local vars are being used 

### ISSUE THREAD

**Issue by deepakjois** [golang/lint#17]
```
package main
import (
    "fmt"
)

func main() {
    res := get(3)
    fmt.Println(res)
}

func get(i int) int {
    if f := i * 2; f == 6 {
        return f
    } else {
        return f * 3
    }
}
```

Running golint displays:

```
main.go:15:9: if block ends with a return statement, so drop this else and outdent its block
```

**Comment by jgroeneveld** [comment]
This is actually a nice pattern to retrieve a value, evaluate the error, return on error or assign the value if there is no error while protecting the namespace.

I see why this is rule exists in the usual cases but when using a local var I was surprised that golint is reporting it.

```
[...] some code
	if loc, err := store.Location(); err != nil {
		return nil, false
	} else {
		doc.Location = loc.ToArray()
	}

// at this point there is no var loc anymore, its only used in the else case
[...] some code
```

**Comment by jriquelme** [comment]
Same opinion as @jgroeneveld here, I have many complains of golink for things like:

```
if nid, err := att(node, "nid"); err != nil {
	return nil, err
} else {
	deliveryMsg.Nid = parseNid(nid)
}
```

I prefer this form for this particular case (doesn't pollute the scope with "temporary" variables).

**Comment by dsymonds** [comment]
The point of a common style is that it overrides personal preference.

**Comment by yurishkuro** [comment]
@dsymonds However, golint readme also says:
> If you find an established style that is frequently violated, and which you think golint could statically check, file an issue.

**Comment by gordallott** [comment]
for anyone coming to this via google search and finding this dead end, i forked and removed this poor decision. just go get that version instead and enjoy less scope leaking 

https://github.com/gordallott/lint

**Comment by delsvr** [comment]
`The point of a common style is that it overrides personal preference.`

This isn't just a question of style, though. These examples are functionally different. The snippets by @jgroeneveld  and @jriquelme limit the scope of variables to the `if` block. The linter enforces polluting the parent scope.

At any rate, there is not even any indication that your example is the "usual Go style". Just see Go's own tutorial on flow control: https://tour.golang.org/flowcontrol/7.

Please reconsider.

**Comment by jkassis** [comment]
come on.

**Comment by siraj-mansour-deltatre** [comment]
This is quite noisy and promotes scope leaking, i am highly for re-opening this issue and changing the rules.
